### PR TITLE
Use AdminClient for topic creation and config updates.

### PR DESCRIPTION
**Pull request by efeg**
Addresses the issue #1096.

### CODE REVIEW

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/KafkaSampleStore.java [hunk 1]
```diff
@@ -168,6 +168,56 @@ public void configure(Map<String, ?> config) {
...
+      } catch (InterruptedException | ExecutionException | TimeoutException e) {
+        LOG.warn("Auto creation of sample store topics failed due to failure to describe cluster.", e);
+        return null;
+      }
+      if (numberOfBrokersInCluster <= 1) {
+        throw new IllegalStateException(String.format("Kafka cluster has less than 2 brokers (brokers in cluster=%d, zookeeper.connect=%s)",
+                                                      numberOfBrokersInCluster, config.get(ExecutorConfig.ZOOKEEPER_CONNECT_CONFIG)));
+      }
+
+      return (short) Math.min(DEFAULT_SAMPLE_STORE_TOPIC_REPLICATION_FACTOR, numberOfBrokersInCluster);
```
Comment (kidkun) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/KafkaSampleStore.java:193]: if we reaches here `numberOfBrokersInCluster` is guaranteed to be equal or larger than `DEFAULT_SAMPLE_STORE_TOPIC_REPLICATION_FACTOR`(which is 2), looks the  `Math.min()` is redundant here.

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/KafkaSampleStore.java [hunk 2]
```diff
@@ -168,6 +168,56 @@ public void configure(Map<String, ?> config) {
...
+    try {
+      CreateTopicsResult createTopicsResult = adminClient.createTopics(Collections.singletonList(topicToBeCreated));
+      createTopicsResult.values().get(topicToBeCreated.name()).get(CruiseControlMetricsUtils.CLIENT_REQUEST_TIMEOUT_MS,
+                                                                   TimeUnit.MILLISECONDS);
+      LOG.info("Topic {} has been created.", topicToBeCreated.name());
+    } catch (InterruptedException | ExecutionException | TimeoutException e) {
+      if (e.getCause() instanceof TopicExistsException) {
+        return false;
+      }
+      LOG.warn("Unable to create topic {}.", topicToBeCreated.name(), e);
```
Comment (kidkun) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/KafkaSampleStore.java:216]: If we are unable to create topic here, shouldn't we throw out exception? I think if this topic does not exist and is fail to be created,  `KafkaSampleStore` is unable to work.

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/KafkaSampleStore.java [hunk 3]
```diff
@@ -215,16 +265,11 @@ public void configure(Map<String, ?> config) {
...
-    boolean zkSecurityEnabled = (Boolean) config.get(ExecutorConfig.ZOOKEEPER_SECURITY_ENABLED_CONFIG);
-    KafkaZkClient kafkaZkClient = KafkaCruiseControlUtils.createKafkaZkClient(connectString,
-                                                                              "KafkaSampleStore",
-                                                                              "EnsureTopicCreated",
-                                                                              zkSecurityEnabled);
-    AdminZkClient adminZkClient = new AdminZkClient(kafkaZkClient);
     AdminClient adminClient = KafkaCruiseControlUtils.createAdminClient((Map<String, Object>) config);
     try {
-      Map<String, List<PartitionInfo>> topics = _consumers.get(0).listTopics();
+      short replicationFactor = sampleStoreTopicReplicationFactor(config, adminClient);
```
Comment (kidkun) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/KafkaSampleStore.java:270]: Looks like `sampleStoreTopicReplicationFactor(config, adminClient)` could return null here, will this be an issue in unboxing?

File: cruise-control-metrics-reporter/src/main/java/com/linkedin/kafka/cruisecontrol/metricsreporter/CruiseControlMetricsReporter.java [hunk 4]
```diff
@@ -220,18 +217,19 @@ protected void maybeUpdateTopicConfig() {
   }
 
   protected void maybeIncreaseTopicPartitionCount() {
+    String cruiseControlMetricsTopic = _metricsTopic.name();
```
[the remark this answers is not in the document]
Comment (efeg) [reply cruise-control-metrics-reporter/src/main/java/com/linkedin/kafka/cruisecontrol/metricsreporter/CruiseControlMetricsReporter.java:220]: Just for making implementations of this function and the `maybeIncreasePartitionCount` in `SamplingUtils` more consistent with each other.  Ideally, a future patch should create a module to combine such common util logic to be used everywhere else. -- having different implementations for the same functionality is error-prone.

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/KafkaSampleStore.java (hunk 1 above)
Comment (efeg) [reply cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/KafkaSampleStore.java:193]: Doesn't this assume that `DEFAULT_SAMPLE_STORE_TOPIC_REPLICATION_FACTOR` will always be `2`? -- i.e. would it still be correct if we change the default RF to `3` or make it configurable in the future?

File: cruise-control-metrics-reporter/src/main/java/com/linkedin/kafka/cruisecontrol/metricsreporter/CruiseControlMetricsReporter.java [hunk 5]
```diff
@@ -182,15 +183,12 @@ protected void createCruiseControlMetricsTopic() throws TopicExistsException {
     try {
       CreateTopicsResult createTopicsResult = _adminClient.createTopics(Collections.singletonList(_metricsTopic));
       createTopicsResult.values().get(_metricsTopic.name()).get(CLIENT_REQUEST_TIMEOUT_MS, TimeUnit.MILLISECONDS);
-      LOG.info("Cruise Control metrics topic {} is created.", _cruiseControlMetricsTopic);
-    } catch (ExecutionException e) {
+      LOG.info("Cruise Control metrics topic {} is created.", _metricsTopic.name());
+    } catch (InterruptedException | ExecutionException | TimeoutException e) {
       if (e.getCause() instanceof TopicExistsException) {
         throw new TopicExistsException(e.getMessage());
```
Comment (viktorsomogyi) [inline cruise-control-metrics-reporter/src/main/java/com/linkedin/kafka/cruisecontrol/metricsreporter/CruiseControlMetricsReporter.java:189]: Why don't you just rethrow `e.getCause()`?

**Pull request merged**